rest_specification/restful_service_specification.md

```diff
@@ -0,0 +1,240 @@
+entity must include a version.
+
+The base URI for a given resource type is based on the *singular* name of the
+resource. Example:
+
+    https://example.com/rest/v1/user
+
+## Operations Against a Resource Type
+
+The following operations may be supported:
+
+| HTTP Operation | Description            | Query Parameters              | Request Body | Response Body  |
+|----------------|------------------------|-------------------------------|--------------|----------------|
+| `POST`         | Create new resource(s) | -                             | JSON array   | -              |
+| `PUT`          | Replace resource(s)    | -                             | JSON array   | -              |
+| `PATCH`        | Update Resource(s)     | -                             | JSON array   | -              |
+| `GET`          | List resource(s)       | `fields`,`page`,`size`,`sort` | -            | paginated list |
+| `DELETE`       | Delete ALL resources   | -                             | -            | -              |
+
+Endpoints may not support all methods (i.e. a service may provide read-only
+access to a resource, or a service may not support certain operations for
+performance or other considerations; for example, `GET` may not be supported
+for `user` if listing all users is disallowed for performance reasons.
+
+## Operations Against Specific Resources
+Specific resources will be addressed through the base URI plus the identifier
+for that resource.
+
+    /user/{id}
+
+The following operations may be supported:
+
+| HTTP Operation | Description                   | Query Parameters | Request Body | Response Body |
+|----------------|-------------------------------|------------------|--------------|---------------|
+| `GET`          | Retrieve the resource         | `fields`         | -            | JSON object   |
+| `PUT`          | Replace the existing resource | -                | JSON object  | -             |
+| `PATCH`        | Update an existing resource   | -                | JSON object  | -             |
+| `DELETE`       | Delete the resource           | -                | -            | -             |
+
+Clients should use the `fields` parameter for better performance.
+
+Implementors should consider breaking out field represented by a JSON array
+into sub-entities if there are performance concerns around its size. This
+consideration should happen regardless of how the data is stored.
+
+## Pagination
+
+Paging shall be implemented via `page` and `size` query parameters. If none are
+provided, the default page size shall be used.
+
+The default page-size is implementation-specific.
+
+| Parameter | Description               |
+|-----------|---------------------------|
+| `page`    | page index, starting at 0 |
+| `size`    | page size                 |
+
+## Sorting
+
+Sorting shall be provided via the `sort` query parameter. `sort` may be
+specified more than once in order to sort by more than one field. `sort` values
```

Comment:
That's a bit surprising. I wonder how widely supported multiple params are while retaining order? Is Java enough for us? From [this](https://www.owasp.org/images/b/ba/AppsecEU09_CarettoniDiPaola_v0.8.pdf) (page 9) it looks like many frameworks do different things when they encounter the same param multiple times, and from [this](http://stackoverflow.com/questions/1746507/authoritative-position-of-duplicate-http-get-query-keys) it sounds like there is no spec to be the decider. I'm getting the impression it would be safer to maybe do something like `sort=foo:asc,bar:desc` (specific syntax aside). Although, if we only care about Java this is probably moot, but personally I was hoping this could be useful beyond Java server apps (which is not to say that multiple of same param excludes them, but I guess it might get complicated).